cmd/ltkeycloak/main.go

```diff
@@ -0,0 +1,468 @@
+	}
+
+	keycloakRealm, err := cmd.Flags().GetString("keycloak-realm")
+	if err != nil {
+		return fmt.Errorf("failed to read flag: %w", err)
+	}
+
+	userPassword, err := cmd.Flags().GetString("set-user-password-to")
+	if err != nil {
+		return fmt.Errorf("failed to read flag: %w", err)
+	}
+
+	keycloakClient := gocloak.NewClient(keycloakHost)
+	ctx := context.Background()
+
+	keycloakToken, err := keycloakClient.LoginAdmin(
+		ctx,
+		deploymentConfig.ExternalAuthProviderSettings.KeycloakAdminUser,
+		deploymentConfig.ExternalAuthProviderSettings.KeycloakAdminPassword,
+		"master", // TODO: Allow specifying the master realm
+	)
+	if err != nil {
+		return fmt.Errorf("failed to refresh keycloak token: %w", err)
+	}
+
+	usersChan := make(chan *model.User, 10000)
+	usersTxtChan := make(chan string, 10000)
+	doneChan := make(chan struct{})
+	workers := 4
+
+	wg := sync.WaitGroup{}
+
+	// Workers to run the sync and call the MM and Keycloak APIs
+	for i := 0; i < workers; i++ {
+		workerNumber := i + 1
+		go func() {
+			refreshTokenTicker := time.NewTicker(30 * time.Second)
+			defer refreshTokenTicker.Stop()
+
+			for {
+				select {
+				case user := <-usersChan:
+					kcUserID, err := keycloakClient.CreateUser(ctx, keycloakToken.AccessToken, keycloakRealm, gocloak.User{
+						Username:      &user.Username,
+						Email:         &user.Email,
+						Enabled:       gocloak.BoolP(true),
+						EmailVerified: &user.EmailVerified,
+						Credentials: &[]gocloak.CredentialRepresentation{
+							{
+								Temporary: gocloak.BoolP(false),
+								Type:      gocloak.StringP("password"),
+								Value:     gocloak.StringP(userPassword),
+							},
+						},
+					})
+					if err != nil {
+						// Ignore already existing users
+						if apiErr, ok := err.(*gocloak.APIError); ok && apiErr.Code == 409 {
+							continue
+						}
+
+						slog.Error("failed to create user in keycloak", slog.String("err", err.Error()))
+						continue
+					}
+
+					user.AuthData = model.NewString(kcUserID)
+					user.AuthService = model.UserAuthServiceSaml
+					user.Password = ""
+
+					_, _, err = mmClient.UpdateUser(ctx, user)
+					if err != nil {
+						slog.Error("failed to update user in mattermost", slog.String("err", err.Error()))
+						continue
+					}
+					slog.Info("migrated user", slog.String("username", user.Username))
+					wg.Done()
+
+				case <-refreshTokenTicker.C:
+					slog.Info("refreshing keycloak token", slog.Int("worker", workerNumber))
+					keycloakToken, err = keycloakClient.LoginAdmin(
+						ctx,
+						deploymentConfig.ExternalAuthProviderSettings.KeycloakAdminUser,
+						deploymentConfig.ExternalAuthProviderSettings.KeycloakAdminPassword,
+						"master", // TODO: Allow specifying the master realm
+					)
+					if err != nil {
+						slog.Error("failed to refresh keycloak token", slog.String("err", err.Error()))
+						close(doneChan)
+						panic(err)
+					}
+
+				case <-doneChan:
```

Comment:
There is a bug here. Actually 2.

1. Closing a channel won't really wait for the `return` statement to execute. So it could be that you call `close(doneChan)` and move on, but this goroutine is yet to return. 
2. You are using a buffered chan, which means that the sender side would be complete, but the consumer might still be waiting to drain the channel. So it would be incorrect to return before draining a channel.

Fixing #1 is easy. Just add yet another waitgroup, or a signalling channel to indicate that the goroutine has finished. See: https://github.com/mattermost/mattermost/blob/e666f7ccfc3575dbe9507c7c68533f717355e0c8/server/channels/app/platform/web_conn.go#L293-L294.

Fixing #2 is complicated. You have to keep reading from the channel until it's drained. I'd suggest to avoid buffered channel altogether and just keep it a simple non-buffered channel.